### PR TITLE
feat(gatsby): use webpack fs caching in prod behind feature flag

### PR DESCRIPTION
**Pull request by pieh**
## Description

This enables webpack 5 persisting caching behind feature flag (reusing existing `PRESERVE_WEBPACK_CACHE`).

Couple of notes:

Cache persistance happen when compiler become idle (i.e. after compilation finished). This needed us to explictely close and wait for compiler to close as there is cache persisting specific hooks. (closing compiler is actually change that we should've done when migrated to webpack@5 anyway - https://webpack.js.org/blog/2020-10-10-webpack-5-release/#compiler-idle-and-close )

On small sites waiting for cache to get persisted can increase overall build times for cold builds. On larger sites it mostly have no visible time effect as persisting cache overlap with html generation (runs concurrently).

On warm builds there is quite a large benefit to using this cache. On very small site browser bundle compilation goes from 5s to <1s. On larger sites from ~60s to ~3s.

[ch27590]

### CODE REVIEW

File: packages/gatsby/src/utils/webpack.config.js [hunk 1]
```diff
@@ -763,6 +763,38 @@ module.exports = async (
     }
   }
 
+  if (stage === `build-javascript` || stage === `build-html`) {
+    const cacheLocation = path.join(
+      program.directory,
```
Comment (KyleAMathews) [inline packages/gatsby/src/utils/webpack.config.js:768]: Could we add these directories as ones that are excluded from the automatic cache clearing?

**Review by KyleAMathews — COMMENTED** [review]
Excited to get this in!

File: packages/gatsby/src/utils/webpack.config.js (hunk 1 above)
Comment (pieh) [reply packages/gatsby/src/utils/webpack.config.js:768]: They fall into what `PRESERVE_WEBPACK_CACHE` already does preserve:
https://github.com/gatsbyjs/gatsby/blob/257207696fec45c4d1bed2f55644d8389f6a9ad6/packages/gatsby/src/services/initialize.ts#L422-L425

So I would say we could work toward making that flag stable and default

**Comment by pieh** [comment]
We should wait for https://github.com/webpack/webpack/pull/13157 to be released (PR was merged already) and bump our `webpack` as it tackles https://github.com/webpack/webpack/issues/13148 which we currently hit.

File: integration-tests/artifacts/gatsby-ssr.js [hunk 2]
```diff
@@ -10,3 +10,23 @@ export function onRenderBody({ setHeadComponents }) {
...
+export function wrapRootElement({ element }) {
+  return (
+    <>
+      <div>
+        (to match src/components/github.js structure without using same imports)
+      </div>
+      {element}
+    </>
+  )
+}
```
Comment (pieh) [inline integration-tests/artifacts/gatsby-ssr.js:32]: this is just to make sure hydration works (even tho text itself doesn't match, html elements structure does). it doesn't matter for tests we do (as we don't do something like cypress in here), but for manual checking it's useful for hydration to actually render things properly

File: packages/gatsby/src/commands/build-javascript.ts [hunk 3]
```diff
@@ -20,7 +28,32 @@ export const buildProductionBundle = async (
...
 
+    const THIRTY_SECONDS = 30 * 1000
+    let cancelCacheNotice: CancelExperimentNoticeCallbackOrUndefined
+    if (
+      process.env.gatsby_executing_command === `build` &&
+      !process.env.GATSBY_EXPERIMENTAL_PRESERVE_WEBPACK_CACHE
+    ) {
+      cancelCacheNotice = showExperimentNoticeAfterTimeout(
+        `Persistent webpack caching`,
+        `https://github.com/gatsbyjs/gatsby/discussions/28331`,
```
Comment (pieh) [inline packages/gatsby/src/commands/build-javascript.ts:39]: this is pointing to existing discussion about cache clearing - not sure if discussion should be edited or new one created for persistent caching

File: packages/gatsby/src/utils/flags.ts [hunk 4]
```diff
@@ -142,7 +142,7 @@ const activeFlags: Array<IFlag> = [
     command: `all`,
     telemetryId: `PreserveWebpackCache`,
     experimental: false,
-    description: `Don't delete webpack's cache when changing gatsby-node.js & gatsby-config.js files.`,
+    description: `Use webpack's persistent caching and don't delete webpack's cache when changing gatsby-node.js & gatsby-config.js files.`,
     umbrellaIssue: `https://gatsby.dev/cache-clearing-feedback`,
```
Comment (pieh) [inline packages/gatsby/src/utils/flags.ts:146]: probably need new shortlink (and possibly new discussion)

Comment (wardpeet) [reply packages/gatsby/src/utils/flags.ts:146]: I'm feeling okay with keeping it as is :shrug:

**Review by wardpeet — DISMISSED** [review]
This looks great!

I have some follow ups:
<details>
<summary>extract webpack compilation code into it's own util</summary>

```
 const compiler = webpack(compilerConfig)
      compiler.run((err, stats) => {
        let activity
        if (process.env.GATSBY_EXPERIMENTAL_PRESERVE_WEBPACK_CACHE) {
          activity = reporter.activityTimer(
            `Caching HTML renderer compilation`,
            { parentSpan }
          )
          activity.start()
        }

        const waitForCompilerClose = new Promise<void>((resolve, reject) => {
          compiler.close(error => {
            if (activity) {
              activity.end()
            }

            if (error) {
              return reject(error)
            }
            return resolve()
          })
        })

        if (err) {
          return reject(err)
        } else {
          return resolve({ stats, waitForCompilerClose })
        }
      })
```
</details>
<details>
<summary>Test that caching works with renaming extensions</summary>

Rename a .js file to .ts

https://app.clubhouse.io/gatsbyjs/story/29635/changing-file-extension-from-js-to-ts-breaks-the-build
</details>

**Comment by pieh** [comment]
Reverted 50% opt-in due to caching issue when changing file extension (i.e. `.jsx` => `.tsx`). So this is just behind feature flag that users can opt into. At least until file extension change is covered. From my testing it seems like https://github.com/gatsbyjs/gatsby/pull/30435 accidentally fixes this (as our current shadowing webpack plugin seems to introduce caching weirdness / doesn't play nice).

**Comment by mjameswh** [comment]
> At least until file extension change is covered. From my testing it seems like #30435 accidentally fixes this (as our current shadowing webpack plugin seems to introduce caching weirdness / doesn't play nice).

Actually, there will some more work required on the theme shadowing plugin to make it safe to play with Webpack 5's cache, notably:
- Use Webpack's file system proxy rather than the `fs` package (should be trivial, but who knows)
- Add all directory and file paths that were tried to the `missingDependencies` collection

I'm afraid that without these changes, many cases where files are added/renamed/deleted will go unnoticed. Fortunately, as far as I can see at this point, that should not be too difficult to get right.

**Pull request merged**